osquery/tables/system/windows/certificates.cpp

```diff
@@ -307,6 +311,230 @@ void parseSystemStoreString(LPCWSTR sysStoreW,
   }
 }
 
+#pragma pack(push, 1)
+struct Header {
+  DWORD propid;
+  DWORD unknown;
+  DWORD size;
+};
+#pragma pack(pop)
+
+Status getEncodedCert(std::basic_istream<BYTE>& blob,
+                      std::vector<BYTE>& encodedCert) {
+  // See these links for details on this magic number:
+  // https://itsme.home.xs4all.nl/projects/xda/smartphone-certificates.html
+  // https://github.com/wine-mirror/wine/blob/f9301c2b66450a1cdd986e9052fcaa76535ba8b7/dlls/crypt32/crypt32_private.h#L146
+  static const DWORD CERT_CERT_PROP_ID = 0x20;
+
+  Header hdr;
+
+  while (true) {
```

Comment:
Yeah, I also think `while(true)` is a slight code smell, but in this particular case I found it to be one of the cleaner solutions I could think of (although I hadn't considered recursion). 

re: infinite loops, those shouldn't be possible because of how we are reading from the stream. So at worst case, we will loop until we reach the end of the stream, then one of the `!blob.good()` checks will fail, and we will exit. (these file streams are typically less than 1kb, fwiw).

The `continue` will hit pretty often. The structure of the file being parsed is basically an array of something like

```
struct Chunk {
  Header hdr;
  char buf[hdr.size]; // variable length depending on the header
};
```

and generally the chunk we're interested in is at the end (so we are `continue`ing past all the other chunks that don't have the `propid` we are looking for). (it would probably be good to have a short description of the format in a comment, I can add one)

I thought of rewriting it using normal loop constructs, or do-while, but it's tricky to avoid duplicated code because of all the error checking on the stream.

A recursive version is doable! It would look something like this:

```
int getEncodedCert(std::basic_istream<char> &blob,
        std::vector<char>& encodedCert) {

    const unsigned long CERT_CERT_PROP_ID = 0x20;

    Header hdr;

    blob.read(reinterpret_cast<char*>(&hdr), sizeof(hdr));
    if (!blob.good()) {
      return Status::failure();
    }

    if (hdr.propid != CERT_CERT_PROP_ID) {
      blob.ignore(hdr.len);
      if (!blob.good()) {
        return  Status::failure();
      }
      return getEncodedCert(blob, encodedCert); // recurse
    }

    encodedCert.resize(hdr.len);
    blob.read(encodedCert.data(), hdr.len);
    if (!blob.good()) {
      return Status::failure();
    }

    return Status::success();
}
```

For me this is slightly less readable than the current (due to the recursion, maybe just me), and since I don't think the infinite loop should be an issue, I'd vote to keep it iterative for the typical reasons we prefer it in C++.

Reply:
I'd so follow your gut on this, if you think the current setup is good I'll be for it :) I tend to push back any time I see a `while(true)`, but as you mention it should for the most part be ok. While a recursive solution might _look_ a little better, it will still have the same consequences for someone trying to exhaust resources. It should be fine to leave.

Reply:
Yeah, I think keeping it as is is my preference. I'd think that a recursive solution might possibly be more error prone due to being able to run out of stack space if there happened to be a blob file with a pathological number of chunks.